design/ingressroute-design.md

```diff
@@ -0,0 +1,398 @@
+This also implies that the IngressRoute root and the tls Secret must live in the same namespace.
+However as mentioned above, the entire routespace (/ onwards) can be delegated to another namespace, which allows operators to define virtual hosts and their TLS configuration in one namespace, and delegate the operation those virtual hosts to another namespace.
+
+# Alternatives Considered
+
+This proposal presents itself as an alternative to more traditional designs that split ingress into two classes of objects; one describing the properties of the ingress virtual host, the other describing the properties of the routes attached to that virtual host.
+
+The downside of these proposals is the minimum use case, the http://hello.world case, needs to two CRD objects per ingress, one stating the url of the virtual host, the other stating the name of the CRD that holds the url of the virtual host.
+
+Restricting who can create this pair of CRD objects further complicates things and draws towards a design with a third and possibly fourth CRD to apply policy to their counterparts.
+
+By "overloading", or as Joe likes to say, making the CRD polymorphic, creates a more complex mental model for complicated deployments, but in return "scales down" to a single CRD containing both the vhost details and the route details (as there is no delegation in the hello world example).
+This property makes the design proposed appealing from a usability standpoint as _most_ ingress usecases are simple--publish my web app on this url--so it feels right to favour a design that does not penalise the default, simple, use case.
+
+- links to other proposals
+
+# Future work
+
+Future work outside the scope of this design includes:
+
+- Limit the set of namespaces where root IngressRoutes are valid. Only those permitted to operate in those namespaces can therefore create virtual hosts and delegate the permission to operate on them to other namespaces. This would most likely be acomplished with a command line flag or ConfigMap.
+- Delegation to matching labels, rather than names, may be added in the future. This is valid, as long as none of the matching IngressRoute objects are roots, because routes are a set, so can be merged from several objects _in the same namespace_.
+
+# Security Concerns
+
+As it relates to the CRD the security implications of this proposal, the model of delegating a part of a vhost's route space to a CRD in another namespace is implies a level of trust.
+I, the owner of the virtual host, delegate to you the /foo prefix and everything under it, implies that you cannot operate on this vhost outside /foo, but it also means that whatever you do in /foo; host malware, leak k8s session keys, is something I'm trusting you not to do.
```

Comment:
I think that it is critical to call out that the trust is one way (I think).  The parent needs to trust the child.  The child doesn't need to trust the parent as anyone that has access to a service can proxy.

Reply:
This is a point of concern for me. By explicitly making trust one way then there is no way to prevent the scenario I outlined.